python/test/regression/test_dedup_mma.py

```diff
@@ -0,0 +1,56 @@
+"""When triton is lowered to llvm it no longer deals with tensors and
```

Comment:
I also generally prefer not using expensive HW to run tests but I went with python because lowering to llvm produces op chains between the mask operation and the dot that involves repeated storing-reading for example

```
    %1506 = llvm.mlir.undef : !llvm.struct<(i1, i1)> loc(#loc127)
    %1507 = llvm.insertvalue %1505, %1506[0] : !llvm.struct<(i1, i1)>  loc(#loc127)
    %1508 = llvm.insertvalue %1505, %1507[1] : !llvm.struct<(i1, i1)>  loc(#loc127)
    %1509 = llvm.extractvalue %824[0] : !llvm.struct<(i1, i1)>  loc(#loc127)
    %1510 = llvm.extractvalue %824[1] : !llvm.struct<(i1, i1)>  loc(#loc127)
    %1511 = llvm.extractvalue %1508[0] : !llvm.struct<(i1, i1)>  loc(#loc127)
    %1512 = llvm.extractvalue %1508[1] : !llvm.struct<(i1, i1)>  loc(#loc127)
```

The error is in the particular way these ops feed into each other. The lit tests that I managed to make were basically unreadable so I opted for a higher level approach that also demonstrates the actual problem.

Reply:
interesting, why do we need mask and dot? This is part of elementwise lowering so we should be able to reduce it to a single op lowering?
The fact that we cannot easily test this analysis individually is a bit concerning to me. 
I know this is not your contribution and you are just trying to fix it but just so that I understand we cannot reproduce this problem by just having a single elementwise op lowering with the right attributes set for the analysis to detect that elements are duplicates?

Reply:
So a few things need to come together for this to be a problem:

- elementwise op
- constancy that is large enough for deduplication to kick in  but not so large that it ends up doing the right thing because all threads see the same value.
- There needs to be MmaEncding -> DotOperandEncoding so that the threads don't actually get contiguous values.

The first is ok, no problems there. 

The second is a problem because I don't know how to reproduce the constancy requirement for the input without copy-pasting the code generated from the mask creation code. I can recognize what the code does if I am well caffeinated and I know what I am looking for but I don't think I could debug it.

The third one is a problem for the same reason as the second: it needs to shuffle the values around the threads storing to shmem and reading back and that is quite lengthy.

So maybe someone (me in this case) can sit stitch parts of generated mlir together into a frankentest which will indeed flag any adjacent change in the future, but IMO it will be very hard to diagnose a failure coming from  such a test. I would rather just embrace that it's a problem emerging from the coincidence of 3 disperate concepts and check it with a regression test. On the other hand I think it would take 5 minutes for someone to read this python code and get what it is guarding against and where to put the print statements in order to trouble-shoot the test blowing up.

I don't know, what do you think?

Reply:
Maybe I'm misunderstanding what is happening here, but couldn't we write a lit test that roughly does something like the following (excuse the pseudo-MLIR):

```mlir
// RUN: triton-opt --convert-triton-gpu-to-llvm
#mma = <fill out you mma layout here>
tt.func @test(%in : tensor<16x1xf16, #mma>) -> tensor<16x16xf16, #mma> {
  %bcast = tt.broadcast %in : tensor<16x16xf32, #mma>
  %neg = arith.negf %bcast : tensor<16x16xf32, #mma>
  return %neg : tensor<16x16xf32, #mma>
}
```

If I understand your change correctly, the deduplication analysis before your change would say that each thread only has a single row, which contains duplicated values, and thus each thread should only do a single negation. Which is wrong for MMA layouts, since each thread will actually have multiple partial rows.
The check can then be on the amount of negation instructions we got in the LLVM code - before you change it will be only one, and after it there should be multiple.

Would that work?